store/store.go

```diff
@@ -766,53 +766,77 @@ func (s *SnapUbuntuStoreRepository) SuggestedCurrency() string {
-type BuyRedirect struct {
+// BuyParameters specifies parameters for store purchases.
+type BuyParameters struct {
+	// Required parameters
+	SnapID        string
+	Name          string
+	Channel       string
+	ExpectedPrice float64
+	Currency      string // ISO 4217 code as string
+	Auther        Authenticator
+
+	// Optional paramters
+	BackendID string // The payment backend to use, e.g. "credit_card", "paypal".
+	MethodID  int    // Identifier for a specific payment method, e.g. a particularly credit card or paypal account.
+}
+
+// BuyResult holds information required to complete the purchase when state
+// is InProgress, case in which it requires user interaction to complete.
+type BuyResult struct {
+	State      string
 	RedirectTo string
 	PartnerID  string
 }
 
-// purchaseInstruction encapsulates the data that must be sent in order to make a purchase from the store.
-// The Device ID can be sent in the extended header "X-Device-Id".
-// The Partner ID (e.g. "bq") can be sent in the extended header "X-Partner-Id".
+// purchaseInstruction holds data sent to the store for purchases.
+// X-Device-Id and X-Partner-Id (e.g. "bq") may be sent as headers.
 type purchaseInstruction struct {
 	SnapID    string  `json:"snap_id"`
 	ItemSKU   string  `json:"item_sku,omitempty"`
 	Amount    float64 `json:"amount,omitempty"`
 	Currency  string  `json:"currency,omitempty"`
 	BackendID string  `json:"backend_id,omitempty"`
-	MethodID  int64   `json:"method_id,omitempty"`
+	MethodID  int     `json:"method_id,omitempty"`
 }
 
 type buyError struct {
 	ErrorMessage string `json:"error_message"`
 }
 
-// Buy the specified snap using the default currency and payment method.
+// Buy sends a purchase request for the specified snap.
 // Returns the state of the purchase: Complete, Cancelled, InProgress or Pending.
-func (s *SnapUbuntuStoreRepository) Buy(snapID, channel string, expectedPrice float64, options BuyOptions, auther Authenticator) (string, *BuyRedirect, error) {
-	if auther == nil {
-		return "", nil, fmt.Errorf("cannot make purchase request from store: no authentication credentials provided")
+func (s *SnapUbuntuStoreRepository) Buy(parameters BuyParameters) (*BuyResult, error) {
+	if parameters.SnapID == "" {
+		return nil, &ErrBuyParameterMissing{parameters.Name, parameters.SnapID, "snap ID"}
+	}
+
+	if parameters.Name == "" {
+		return nil, &ErrBuyParameterMissing{parameters.Name, parameters.SnapID, "name"}
+	}
+
+	if parameters.Channel == "" {
+		return nil, &ErrBuyParameterMissing{parameters.Name, parameters.SnapID, "channel"}
+	}
+
+	if parameters.ExpectedPrice <= 0 {
+		return nil, &ErrBuyParameterMissing{parameters.Name, parameters.SnapID, "price"}
+	}
+
+	if parameters.Currency == "" {
+		return nil, &ErrBuyParameterMissing{parameters.Name, parameters.SnapID, "currency"}
+	}
+
+	if parameters.Auther == nil {
+		return nil, &ErrBuyParameterMissing{parameters.Name, parameters.SnapID, "authentication credentials"}
```

Comment:
The main point to have an error type is to treat the error specially in code, and we generally add this either if we're almost certainly sure that we'll need to handle this case specially, or when we are missing it already somewhere else in the code.

These errors above are neither. They are just informational messages that we'll probably never handle specially, since we really cannot do much about them other than letting the user know. As such, plain fmt.Errorf messages would be much better here, with more clarity and less boilerplate on errors.go.

You can also leverage normal code to avoid some of the duplication that exists in those types. For example:
```
prefix := fmt.Sprintf("cannot buy %: ", options.SnapName)
...
if ... {
        return fmt.Errorf(prefix + "authentication credentials missing")
}
```

Doesn't that look a lot simpler and more clear?